### PR TITLE
[xpk] Support base-docker-image and script-dir

### PR DESCRIPTION
**Pull request by Obliviour**
- Adds support to build local files into the docker image
- Adds readme example for how to do this
- Continues to support docker-image for not building local files into the image.

Tested
- Ran examples with local file changes
- Verified that directory path has to be valid
- Verified that invalid group of arguments don't work

### CODE REVIEW

File: xpk/README.md [hunk 1]
```diff
@@ -106,6 +106,45 @@ cleanup with a `Cluster Delete`.
...
+Example with defaults which pulled the local directory into the base image:
+```shell
+echo -e '#!/bin/bash \n echo "Hello world from a test script!"' > test.sh
+python3 xpk/xpk.py workload create --cluster xpk-test \
+--workload xpk-test-workload-base-image --command "bash test.sh" \
+--tpu-type=v5litepod-16 --num-slices=1
+```
+
+## `--docker-image`
+If a user wants to directly set the docker image used, set `--docker-image`
```
Comment (rwitten) [inline xpk/README.md:134]: "If a user wants to directly set the docker image used and not layer in the current working directory"

File: xpk/README.md [hunk 2]
```diff
@@ -106,6 +106,45 @@ cleanup with a `Cluster Delete`.
...
+to the image to use in the workload.
+
+Example:
+```shell
+# (Optional) Cacheimage to enable faster start up times.
```
Comment (rwitten) [inline xpk/README.md:139]: Cacheimage to enable faster start up times, typically only required on jobs running on thousands of accelerators.

File: xpk/README.md (hunk 1 above)
Comment (rwitten) [reply xpk/README.md:134]: Also probably this example should include a `--base-docker-image` as a command line argument?

Comment (rwitten) [reply xpk/README.md:134]: Since that is the way we recommended.

File: xpk/README.md (hunk 2 above)
Comment (rwitten) [reply xpk/README.md:139]: Actually I'm make a note on 137 that is like "Example (recommended flow when running on more than 10k accelerators)" and get rid of the lower one.

File: xpk/README.md [hunk 3]
```diff
@@ -106,6 +106,45 @@ cleanup with a `Cluster Delete`.
...
+python3 xpk/xpk.py workload create --cluster xpk-test \
+--workload xpk-test-workload-base-image --command "bash test.sh" \
+--tpu-type=v5litepod-16 --num-slices=1
+```
+
+## `--docker-image`
+If a user wants to directly set the docker image used, set `--docker-image`
+to the image to use in the workload.
+
+Example:
```
Comment (rwitten) [inline xpk/README.md:137]: "Recommended Flow For Normal Sized Jobs (fewer than 10k accelerators)"

File: xpk/xpk.py [hunk 4]
```diff
@@ -1197,6 +1213,84 @@ def validate_docker_image(args) -> int:
...
+        f' You should be able to navigate to the URL {args.base_docker_image}'
+        f' in {args.project}.'
+    )
+    xpk_exit(1)
+
+  # Pick a randomly generated `tag_length` character docker tag.
+  tag_length = 8
+  tag_name = ''.join(random.choices(string.ascii_lowercase, k=tag_length))
+
+  cloud_docker_image = f'gcr.io/{args.project}/{docker_name}:{tag_name}'
```
Comment (rwitten) [inline xpk/xpk.py:1257]: can we include a datetime as well?

**Review by rwitten — CHANGES_REQUESTED** [review]
Documentation nits but total fire!!!!

File: xpk/README.md (hunk 1 above)
Comment (Obliviour) [reply xpk/README.md:134]: I added --base-docker-image to the example in the `--base-docker-image section`. LMK if that is what you intended.

File: xpk/README.md (hunk 2 above)
Comment (Obliviour) [reply xpk/README.md:139]: Done.

File: xpk/README.md (hunk 3 above)
Comment (Obliviour) [reply xpk/README.md:137]: I added this section and took out the cacheimage step in this example.

File: xpk/xpk.py (hunk 4 above)
Comment (Obliviour) [reply xpk/xpk.py:1258]: Done

**Review by Obliviour — COMMENTED** [review]
Done

File: xpk/README.md [hunk 5]
```diff
@@ -115,3 +162,29 @@ feedback.
 * Workload create accepts a --env-file flag to allow specifying the container's
 environment from a file. Usage is the same as Docker's
 [--env-file flag](https://docs.docker.com/engine/reference/commandline/run/#env)
+
+* Access stable fleet TPU machines by adding `--host-maintenance-interval=PERIODIC`
```
Comment (rwitten) [inline xpk/README.md:166]: I don't think StableFleet is GA so I think we should remove this.

File: xpk/README.md [hunk 6]
```diff
@@ -106,6 +106,53 @@ cleanup with a `Cluster Delete`.
...
+--workload xpk-test-workload-base-image --command "bash test.sh" \
+--tpu-type=v5litepod-16 --num-slices=1 --base-docker-image=python:3.10
+```
+
+## Optional Direct Docker Image Configuration: `--docker-image`
+If a user wants to directly set the docker image used and not layer in the
+current working directory, set `--docker-image` to the image to be use in the
+workload.
+
+* Recommended Flow For Normal Sized Jobs (fewer than 10k accelerators)
```
Comment (rwitten) [inline xpk/README.md:139]: Sorry for sending you on a wild goose chase -- but this isn't really the recommended flow less than 10k accelerators. The recommended flow for small jobs is the autolayered approach. Can we call this "Running with ---docker-image"

File: xpk/README.md [hunk 7]
```diff
@@ -106,6 +106,53 @@ cleanup with a `Cluster Delete`.
...
+  ```shell
+  python3 xpk/xpk.py workload create --cluster xpk-test \
+  --workload xpk-test-workload-base-image --command "bash test.sh" \
+  --tpu-type=v5litepod-16 --num-slices=1 --docker-image=gcr.io/your_docker_image
+  ```
+
+* Recommended Flow For Large Sized Jobs (more than 10k accelerators):
```
Comment (rwitten) [inline xpk/README.md:146]: And this one is perfect -- because this is recommended!

File: xpk/README.md [hunk 8]
```diff
@@ -106,6 +106,53 @@ cleanup with a `Cluster Delete`.
...
+
+## Recommended / Default Docker Flow: `--base-docker-image` and `--script-dir`
+Pull in a local directory with scripts or files interested in running in
+`xpk workload create`.
+
+* `--base-docker-image` to set the base image that xpk will start with.
+* `--script-dir` to set which directory to pull into the image.
+* See `python3 xpk/xpk.py workload create --help` for more.
+
+Example with defaults which pulled the local directory into the base image:
```
Comment (rwitten) [inline xpk/README.md:125]: Can we add a second one that has --base-docker-image=gcr.io/your_dependencies_docker_image and say that is the recommended flow for <10k accelerators?

**Review by rwitten — CHANGES_REQUESTED** [review]
My feedback last time was unclear. Hopefully this helps. Also happy to pair program this over the line.

File: xpk/README.md (hunk 5 above)
Comment (Obliviour) [reply xpk/README.md:166]: > I don't think StableFleet is GA so I think we should remove this.

Done.

File: xpk/README.md (hunk 6 above)
Comment (Obliviour) [reply xpk/README.md:139]: Done, and now understood!

File: xpk/README.md (hunk 8 above)
Comment (Obliviour) [reply xpk/README.md:125]: Done!

**Comment by Obliviour** [comment]
> My feedback last time was unclear. Hopefully this helps. Also happy to pair program this over the line.

Thanks Rafi! I think it makes sense to me now. It always helps!

File: xpk/README.md [hunk 9]
```diff
@@ -141,6 +141,62 @@ cleanup with a `Cluster Delete`.
...
+  --workload xpk-test-workload-base-image --command "bash test.sh" \
+  --tpu-type=v5litepod-16 --num-slices=1 --base-docker-image=python:3.10
+  ```
+
+* Recommended Flow For Normal Sized Jobs (fewer than 10k accelerators):
+  ```shell
+  python3 xpk/xpk.py workload create --cluster xpk-test \
+  --workload xpk-test-workload-base-image --command "bash custom_script.sh" \
+  --base-docker-image=gcr.io/your_dependencies_docker_image \
+  --script-dir=/custom_dir/
```
Comment (rwitten) [inline xpk/README.md:174]: no need to set this! it is the default.

File: xpk/README.md [hunk 10]
```diff
@@ -141,6 +141,62 @@ cleanup with a `Cluster Delete`.
...
+See below for the two ways that xpk supports configuring docker images for the
+workload. Do not mix arguments from these two ways like `--base-docker-image`
+and `--docker-image` in the same command.
+
+## Recommended / Default Docker Flow: `--base-docker-image` and `--script-dir`
+Pull in a local directory with scripts or files interested in running in
+`xpk workload create`.
+
+* `--base-docker-image` to set the base image that xpk will start with.
+* `--script-dir` to set which directory to pull into the image.
```
Comment (rwitten) [inline xpk/README.md:157]: we hsould make clear we don't expect to set this typically

**Review by rwitten — APPROVED** [review]
Tiny nits but approving.

**Pull request merged**